controllers/grid/queries/QueriesGridHandler.inc.php

```diff
@@ -520,6 +560,47 @@ function updateQuery($args, $request) {
 			)
 		);
 	}
+
+	/**
+	 * Leave query
+	 * @param $args array
+	 * @param $request PKPRequest
+	 * @return JSONMessage JSON object
+	 */
+	function leaveQuery($args, $request) {
+		$queryId = $args['queryId'];
+		$user = $request->getUser();
+		$context = $request->getContext();
+		if ($user && $user->hasRole(ROLE_ID_MANAGER, $context->getId()) && $this->_getCanLeave($user->getId(), $queryId)) {
+			$queryDao = DAORegistry::getDAO('QueryDAO');
+			$queryDao->removeParticipant($queryId, $user->getId());
+			$json = new JSONMessage();
+			$json->setEvent('user-left-discussion');
+		} else {
+			$json = new JSONMessage(false);
+		}
+		return $json;
+	}
+
+	/**
+	 * Check if the current user can leave a query. Only allow if query has more than two participants.
+	 * @param $userId int
+	 * @param $queryId int
+	 * @return boolean
+	 */
+	function _getCanLeave($userId, $queryId) {
```

Comment:
Let's call this `_getCurrentUserCanLeave` and rely on the request user roles. It will save us a couple of DB hits. It would look something like this:

```php
function _getCurrentUserCanLeave($queryId) {
  $userRoles = $this->getAuthorizedContextObject(ASSOC_TYPE_USER_ROLES);
  if (!in_array(ROLE_ID_MANAGER, $userRoles)) {
    return false;
  }
  $queryDao = DAORegistry::getDAO('QueryDAO');
  $participantIds = $queryDao->getParticipantIds($queryId);
  if (count($participantIds) < 3) {
    return false;
  }
  $user = Application::get()->getRequest()->getUser();
  return in_array($user->getId(), $participantIds);
}
```

That will make your conditional statements above a little bit clearer. And whenever we call `$user->hasRole()` it does another hit on the database. No big deal but I'm trying to phase out its use (it was my mistake to implement it).

Reply:
check